### PR TITLE
[WIP] Better stderr testing in functional tests

### PR DESCRIPTION
**Pull request by jnewbery**
(WIP because if we're not careful this could result in a lot of false +ves. Should be tested on different platforms to verify portability).

#12362 was only observed when running the functional tests locally because:
- by defatul libc logs to `/dev/tty` instead of stderr
- the functional tests only check for substring inclusion in stderr when we're expecting bitcoind to fail.

This PR tightens our checking of stderr and will cause tests to fail if there is any unexpected message in stderr:

- commit _rename TestNode to TestP2PConn in tests_ is not necessary in this PR, but it removes a name conflict in the functional tests (which makes it difficult to see all the places that `TestNode` is being called)
- commit _Move assert_start_raises_init_error method to TestNode_ moves a method from test_framework to test_node (in general, methods which only affect a single node should be methods in `TestNode`).
- commit _Write stdout/stderr to datadir instead of temp file_ writes stderr to a file in the datadir instead of a temporary file. This helps with debugging in the case of failure.
- commit _Use LIBC_FATAL_STDERR_=1 in tests_ ensures that libc failures are logged to stderr instead of the terminal.
- commit _Update assert_start_raises_init_eror() to test for an exact m…_ changes `assert_start_raises_init_eror()` to match on an exact regex instead of a substring (with due [apologies](https://github.com/bitcoin/bitcoin/pull/12302#issuecomment-361691241) to @laanwj for #12302).
- commit _Assert that bitcoind stdout is empty on shutdown_ asserts that stderr is empty on bitcoind shutdown.

### CODE REVIEW

File: test/functional/test_framework/test_node.py [hunk 1]
```diff
@@ -160,6 +173,38 @@ def is_node_stopped(self):
...
+            except Exception as e:
+                assert 'bitcoind exited' in str(e)  # node must have shutdown
+                self.running = False
+                self.process = None
+                # Check stderr for expected message
+                if expected_msg is not None:
+                    log_stderr.seek(0)
+                    stderr = log_stderr.read().decode('utf-8')
+                    if re.fullmatch(expected_msg + '\n', stderr) is None:
+                        raise AssertionError('Expected stdout "{}" does not match stdout:\n"{}'.format(expected_msg, stderr))
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py]: Nit: Should say stderr in the string?

**Review by maflcko — COMMENTED** [review]
The scripted diff should really go into a separate pull

**Comment by laanwj** [comment]
Concept ACK

**Comment by randolf** [comment]
Concept ACK.

**Comment by Sjors** [comment]
I ran `test_runner.py` on MacOS 10.13.3. `wallet_encryption.py`, `wallet_multiwallet.py` and `wallet_multiwallet.py --usecli` failed. I ran them again individually and they still failed. See [logs](https://gist.github.com/Sjors/fcd54278678d989e37f8a478c7e01498). The first test fails on master two, but those other two pass on master (ffc6e48b2).

**Comment by jnewbery** [comment]
I've pulled out the first commit into #12728, and the `assert_start_raises_init_error()` commits have been pulled into #12718.

I'll leave this PR open for now, but will close and replace with a separate PR for the libc/stdout changes once those other two PRs are merged.

**Comment by MarcoFalke** [comment]
Needs rebase

**Comment by jnewbery** [comment]
Parts of this PR have been merged in other PRs. Closing this in favour of #12755 which contains only the parts which have not yet been merged.

**Pull request closed without merging**